### PR TITLE
Fix Issue 22998 - Update to zlib 1.2.12

### PR DESCRIPTION
**Pull request by ibara**
Hello --

I've updated the copy of zlib in Phobos.

Here's a reprinting of the big-ticket items from the zlib.net homepage:
Version 1.2.12 has these key improvements over 1.2.11:

* Fix a deflate bug when using the Z_FIXED strategy that can result in out-of-bound accesses.
* Fix a deflate bug when the window is full in deflate_stored().
* Speed up CRC-32 computations by a factor of 1.5 to 3.
* Use the hardware CRC-32 instruction on ARMv8 processors.
* Speed up crc32_combine() with powers of x tables.
* Add crc32_combine_gen() and crc32_combine_op() for fast combines.
Due to the bug fixes, any installations of 1.2.11 should be replaced with 1.2.12.

Also maintains the fix from here:
https://github.com/ibara/phobos/commit/1b6cf80232f19e581f72869a4b339580416dd72d

Everything appears OK on my OpenBSD/amd64 machine. But perhaps more testing from others is in order.

### CODE REVIEW

**Comment by ibara** [comment]
(Force-pushed to quell a buildkit/phobos error complaining about the style of a comment...)
And also to maybe to get the bot to recognize this PR.

File: etc/c/zlib.d [hunk 1]
```diff
@@ -1751,6 +1756,18 @@ uint crc32_combine(uint crc1, uint crc2, z_off_t len2);
...
+*/
+
+uint crc32_combine_op(uint crc1, uint crc2, uint op);
+/*
+     Give the same result as crc32_combine(), using op in place of len2. op is
+   is generated from len2 by crc32_combine_gen(). This will be faster than
+   crc32_combine() if the generated op is used more than once.
+*/
 
                         /* various hacks, don't look :) */
```
Comment (ibuclaw) [inline etc/c/zlib.d:1772]: 🙈

**Review by ibuclaw — APPROVED** [review]

File: etc/c/zlib.d [hunk 2]
```diff
@@ -566,14 +567,13 @@ int deflateInit2(z_streamp strm,
...
-   fields next_in, zalloc, zfree and opaque must be initialized before by the
-   caller.
+   fields zalloc, zfree and opaque must be initialized before by the caller.
 
      The method parameter is the compression method.  It must be Z_DEFLATED in
    this version of the library.
 
      The windowBits parameter is the base two logarithm of the window size
-   (the size of the history buffer).  It should be in the range 8 .. 15 for this
+   (the size of the history buffer).  It should be in the range 8..15 for this
```
Comment (ibuclaw) [inline etc/c/zlib.d:576]: The style checker will always fire for this, so better revert to the previous range style.

Comment (ibara) [reply etc/c/zlib.d:576]: OK, will do.

Comment (ibara) [reply etc/c/zlib.d]: Done.

**Comment by dkorpel** [comment]
Please add "Fix Issue 22998" to the commit title so the bot links it to the bugzilla issue.

**Review by RazvanN7 — APPROVED** [review]

**Comment by lulcat** [comment]
Hi. I was just looking at the zlib 1.2.11 to 1.2.12 this very moment and saw your post from these days. Great! This was due to wanting to update zstd to current. I believe that finally there has come a natural replacement of the venerable zlib/[gzip]. Anyway, thanks for this, and wondering @ibara , if this is something you might join in or have plans about etc? I will test the 1.2.12, I suppose I need to grab the git version?
On a side note, why is the sources of zlib-1.2.11 in etc/c/zlib? I noticed it was mainly the !_DMD macro thing which was the only difference. Does dmd compile with those sources implicitly or something? (I ask because I am uncertain if I locally should drop zstd lib sources potentially in there or so).

**Comment by ibara** [comment]
Since you pinged me, when you say updating zstd, do you mean the dub package here: https://code.dlang.org/packages/zstd ? If so, it's not my package so I can't tell you when it will be updated.

If you're talking about replacing zlib with zstd in Phobos, then I imagine that would be a bigger change in need of serious discussion (and numbers) to demonstrate that it is a net gain. I am not part of the DLF, just a user and I guess de facto OpenBSD OS maintainer for D, so I doubt a proposal from me would carry much weight without lots of hard numbers. I don't have the time to do that right now.

**Comment by ibuclaw** [comment]
I'm hearing whispers that 1.2.13 is already upcoming due to a [CRC regression](https://github.com/madler/zlib/commit/ec3df00224d4b396e2ac6586ab5d25f673caa4c2) that's been fixed in the devel branch.  If it isn't happening before the next release, perhaps we should cherry-pick?

**Comment by ibuclaw** [comment]
@ibara I've just remembered that a changelog entry would be ideal so it gets picked up in the notes for the next release.

**Comment by ibara** [comment]
> I'm hearing whispers that 1.2.13 is already upcoming due to a [CRC regression](https://github.com/madler/zlib/commit/ec3df00224d4b396e2ac6586ab5d25f673caa4c2) that's been fixed in the devel branch. If it isn't happening before the next release, perhaps we should cherry-pick?

Or I can just update again. It's not too difficult or time-consuming.

**Comment by lulcat** [comment]
Looking at how that bug is worded, cherrypicking might not be uber important (I don't dare say), but had it been literally erronous crc's in current, then yes. Regarding zstd @ibara , yes I mean tthe dub package. I have updated the interface file (using dstep and cleaning up a few bits and bobs there), but I haven't touched the d files the original author had layered on top. It works as before it seems. With that said, in the examples, I get a true vs. false result in 'streaming.d' when doing binary files vs text files. So I assume it's got to do with some string encoding or something although I am not quite getting it why it happens. I'm not tremendous at all with D so. With that said, should someone wish to update it, I can link the dstep file, although it did a fairly good job. Failed to convert two 0ULL to ulong and fixing the version string pretty much as before/zlib.d does.

**Pull request merged**